File: DeepspaceRuby/lib/SpaceStation.rb
#encoding: UTF-8

require_relative 'SpaceStationToUI'
require_relative 'ShotResult'
require_relative 'CardDealer'
require_relative 'Damage'
require_relative 'Transformation'

module Deepspace
    class SpaceStation
        @@MAXFUEL = 100.0
        @@SHIELDLOSSPERUNITSHOT = 0.1

        @ammoPower
        @fuelUnits
        @name
        @nMedals
        @shieldPower
        @pendingDamage
        @hangar
        @weapons
        @shieldBoosters

        def initialize(n,supplies)
            @name = n;
            @ammoPower = supplies.ammoPower
            @fuelUnits = supplies.fuelUnits
            @shieldPower = supplies.shieldPower
            @nMedals = 0
            @pendingDamage = nil
            @hangar = nil
            @weapons = Array.new
            @shieldBoosters = Array.new
        end

        def copy(station)
            @ammoPower = station.ammoPower
            @fuelUnits = assignFuelValue(station.fuelUnits)
            @name = station.name
            @nMedals = station.nMedals
            @shieldPower = station.shieldPower
            @pendingDamage = station.pendingDamage
            @weapons = station.weapons.dup
            @shieldBoosters = station.shieldBoosters.dup
            @hangar = station.hangar
        end

        private
        def assignFuelValue(f)
            if f > @@MAXFUEL
                f = @@MAXFUEL
            end
            @fuelUnits = f
        end

        def cleanPendingDamage
            if @pendingDamage.hasNoEffect
                @pendingDamage = nil
            end
        end

        public

        def cleanUpMountedItems
            @weapons.delete_if{|weapon| weapon.uses == 0}
            @shieldBoosters.delete_if{|sB| sB.uses == 0}
        end

        def discardHangar
            if @hangar != nil
                @hangar = nil
            end
        end

        def discardShieldBooster(i)
            if ((0 <= i) && (i < @shieldBoosters.length()))
                @shieldBoosters.delete_at(i)
                if @pendingDamage != nil
                    @pendingDamage.discardShieldBooster()
                    cleanPendingDamage()
                end
            end
        end

        def discardShieldBoosterInHangar(i)
            if @hangar != nil
                @hangar.removeShieldBooster(i)
            end
        end

        def discardWeapon(i)
            if ((0 <= i) && (i < @weapons.length()))
                weapon = @weapons.delete_at(i)
                if @pendingDamage != nil
                    @pendingDamage.discardWeapon(weapon)
                    cleanPendingDamage()
                end
            end
        end

        def discardWeaponInHangar(i)
            if @hangar != nil
                @hangar.removeWeapon(i)
            end
        end

        def fire
            factor = 1.0
            size = @weapons.length
            for i in (0..size-1)
                w = @weapons.at(i)
                factor = factor * w.useIt()
            end
            return factor*@ammoPower
        end

        def ammoPower
            @ammoPower
        end

        def fuelUnits
            @fuelUnits
        end

        def name
            @name
        end
        
        def nMedals
            @nMedals
        end

        def shieldPower
            @shieldPower
        end
          
        def pendingDamage
            @pendingDamage
        end

        def hangar
            @hangar
        end
          
        def weapons
            @weapons
        end
            
        def shieldBoosters
            @shieldBoosters
        end
           
        def getSpeed
            return @fuelUnits.to_f/@@MAXFUEL.to_f
        end

        def getUIversion
            return SpaceStationToUI.new(self)
        end

        def mountShieldBooster(i)
            if @hangar != nil
                if ((s=@hangar.removeShieldBooster(i)) != nil)
                    @shieldBoosters.push(s)
                end  
            end
        end

        def mountWeapon(i)
            if @hangar != nil
                if ((w=@hangar.removeWeapon(i)) != nil)
                    @weapons.push(w)
                end  
            end
        end

        def move
            @fuelUnits -= (getSpeed*fuelUnits)
            if @fuelUnits < 0
                @fuelUnits = 0
            end
        end

        def protection
            factor = 1.0
            size = @shieldBoosters.length
            for i in (0..size-1)
                s = @shieldBoosters.at(i)
                factor = factor * s.useIt()
            end
            return factor*@shieldPower
        end

        def receiveHangar(h)
            if @hangar == nil
                @hangar = h
            end
        end

        def receiveShieldBooster(s)
            if @hangar == nil
                return false
            end
            return @hangar.addShieldBooster(s)
        end

        def receiveShot(shot)
            myProtection = protection()
            if myProtection >= shot
                @shieldPower -= @@SHIELDLOSSPERUNITSHOT*shot
                @shieldPower = [@shieldPower,0.0].max
                return ShotResult::RESIST
            else
                @shieldPower = 0.0
                return ShotResult::DONOTRESIST
            end

        end

        def receiveSupplies(s)
            @ammoPower += s.ammoPower
            @shieldPower += s.shieldPower
            assignFuelValue(@fuelUnits+s.fuelUnits)
        end

        def receiveWeapon(w)
            if @hangar == nil
                return false
            end
            return @hangar.addWeapon(w)
        end

        def setLoot(l)
            dealer = CardDealer.instance
            h = l.nHangars
            if h>0
                hangar = dealer.nextHangar
                receiveHangar(hangar)
            end

            elements = l.nSupplies
            for i in (0..elements-1)
                sup = dealer.nextSuppliesPackage
                receiveSupplies(sup)
            end

            elements = l.nWeapons
            for i in (0..elements-1)
                w = dealer.nextWeapon
                receiveWeapon(w)
            end

            elements = l.nShields
            for i in (0..elements-1)
                sh = dealer.nextShieldBooster
                receiveShieldBooster(sh)
            end

            medals = l.nMedals
            @nMedals += medals

            if l.efficient
                return Transformation::GETEFFICIENT
            elsif l.spaceCity
                return Transformation::SPACECITY
            else
                return Transformation::NOTRANSFORM
            end
        end

        def setPendingDamage(d)
            @pendingDamage = d.adjust(@weapons,@shieldBoosters)
            cleanPendingDamage
        end

        def validState
            if @pendingDamage == nil
                return true
            end
            return pendingDamage.hasNoEffect            
        end

        def to_s
            namE = "+ NAME: #{@name}\n"
            aP = "+ AMMOPOWER: #{@ammoPower} "
            fU = "+ FUELUNITS: #{@fuelUnits} "
            sP = "+ SHIELDPOWER: #{@shieldPower}\n"
            nMed = "+ NMEDALS: #{@nMedals}\n"
            
            pDam = "+ PENDING DAMAGE: Ninguno \n"
            if pendingDamage != nil
                pDam = "+ PENDING DAMAGE: \n#{@pendingDamage.to_s}"
            end
            
            wP = "+ WEAPONS MOUNTED: Ninguno \n"
            if !weapons.empty?
                wP = "\n+ WEAPONS MOUNTED: \n"
                @weapons.each { |w|
                wP += "#{w.to_s}"
                }
            end
            
            sB = "+ SHIELDBOOSTERS MOUNTED: Ninguno \n"
            if !shieldBoosters.empty?
                sB = "\n+ SHIELDBOOSTERS MOUNTED: \n"
                @shieldBoosters.each { |shieldB|
                sB += "#{shieldB.to_s}"
                }
            end
            
            hG = "+ HANGAR: Ninguno \n"
            if hangar != nil
                hG = "\n+ HANGAR: \n#{@hangar.to_s}\n"
            end
            
            return namE + aP + fU + nMed + sP + pDam + wP + sB + hG
        end
    end
end

File: deepspace/SpaceStation.java
/*
 * @author: Jesús García León
 * @file: SpaceStation
 */
package deepspace;

import java.util.ArrayList;

/**
 * @brief Estación del jugador
 */
public class SpaceStation implements SpaceFighter{

    private static final int MAXFUEL = 100;
    private static final float SHIELDLOSSPERUNITSHOT = 0.1f;

    private float ammoPower;
    private float fuelUnits;
    private String name;
    private int nMedals;
    private Damage pendingDamage;
    private float shieldPower;
    private Hangar hangar;
    private ArrayList<Weapon> weapons = new ArrayList<>();
    private ArrayList<ShieldBooster> shieldBoosters = new ArrayList<>();

    private void assignFuelValue(float f) {
        if (f > MAXFUEL) {
            fuelUnits = MAXFUEL;
        } else {
            fuelUnits = f;
        }
    }

    private void cleanPendingDamage() {
        if (pendingDamage.hasNoEffect()) {
            pendingDamage = null;
        }
    }

    SpaceStation(String n, SuppliesPackage supplies) {
        name = n;
        ammoPower = supplies.getAmmoPower();
        fuelUnits = supplies.getFuelUnits();
        shieldPower = supplies.getShieldPower();
        nMedals = 0;
        pendingDamage = null;
        hangar = null;
    }
    
    SpaceStation(SpaceStation station){
        ammoPower = station.ammoPower;
        fuelUnits = station.fuelUnits;
        name = station.name;
        nMedals = station.nMedals;
        shieldPower = station.shieldPower;
        pendingDamage = station.pendingDamage;
        
        for(Weapon w : station.weapons){
            weapons.add(w);
        }
        for(ShieldBooster s : station.shieldBoosters){
            shieldBoosters.add(s);
        }
        hangar = station.hangar;
    }

    public void cleanUpMountedItems() {
        int i = 0;

        while (i < weapons.size()) {
            if (weapons.get(i).getUses() == 0) {
                weapons.remove(weapons.get(i));
            } else {
                i++;
            }
        }

        i = 0;

        while (i < shieldBoosters.size()) {
            if (shieldBoosters.get(i).getUses() == 0) {
                shieldBoosters.remove(shieldBoosters.get(i));
            } else {
                i++;
            }
        }
    }

    public void DiscardHangar() {
        if (!(hangar == null)) {
            hangar = null;
        }
    }

    public void DiscardShieldBooster(int i) {
        if((0 <= i) && (i < shieldBoosters.size())){
            shieldBoosters.remove(i);
            if(pendingDamage!=null){
                pendingDamage.discardShieldBooster();
                cleanPendingDamage();
            }
        } 
    }

    public void DiscardShieldBoosterInHangar(int i) {
        if (hangar != null) {
            hangar.removeShieldBooster(i);
        }
    }

    public void DiscardWeapon(int i) {
        if((0 <= i) && (i < weapons.size())){
            Weapon weapon = weapons.remove(i);
            if(pendingDamage!=null){
                pendingDamage.discardWeapon(weapon);
                cleanPendingDamage();
            }
        }      
    }

    public void DiscardWeaponInHangar(int i) {
        if (hangar != null) {
            hangar.removeWeapon(i);
        }
    }

    @Override
    public float fire() {
        float factor = 1.0f;
        int size = weapons.size();
        for (int i = 0; i < size; i++){
            Weapon w = weapons.get(i);
            factor *= w.useIt();
        }
        return factor*ammoPower;
    }

    public float getAmmoPower() {
        return ammoPower;
    }

    public float getFuelUnits() {
        return fuelUnits;
    }

    public Hangar getHangar() {
        return hangar;
    }

    public String getName() {
        return name;
    }

    public int getNMedals() {
        return nMedals;
    }

    public Damage getPendingDamage() {
        return pendingDamage;
    }

    public float getShieldPower() {
        return shieldPower;
    }

    public ArrayList<ShieldBooster> getShieldBoosters() {
        return shieldBoosters;
    }

    public float getSpeed() {
        return fuelUnits / MAXFUEL;
    }

    public ArrayList<Weapon> getWeapons() {
        return weapons;
    }

    SpaceStationToUI getUIversion() {
        return new SpaceStationToUI(this);
    }

    public void mountShieldBooster(int i) {
        ShieldBooster s;
        if (hangar != null) {
            if ((s = hangar.removeShieldBooster(i)) != null) {
                shieldBoosters.add(s);
            }
        }
    }

    public void mountWeapon(int i) {
        Weapon w;
        if (hangar != null) {
            if ((w = hangar.removeWeapon(i)) != null) {
                weapons.add(w);
            }
        }
    }

    public void move() {
        fuelUnits -= getSpeed()*fuelUnits;
        if (fuelUnits < 0) {
            fuelUnits = 0;
        }
    }

    @Override
    public float protection() {
        float factor = 1.0f;
        int size = shieldBoosters.size();
        for(int i = 0; i < size; i++){
            ShieldBooster s = shieldBoosters.get(i);
            factor *= s.useIt();
        }
        return factor*shieldPower;
    }

    public void receiveHangar(Hangar h) {
        if (hangar == null) {
            hangar = h;
        }
    }

    public boolean receiveShieldBooster(ShieldBooster s) {
        if (hangar == null) {
            return false;
        }

        return hangar.addShieldBooster(s);
    }

    @Override
    public ShotResult receiveShot(float shot) {
        float myProtection = protection();
        if(myProtection >= shot){
            shieldPower -= SHIELDLOSSPERUNITSHOT*shot;
            shieldPower = Math.max(0.0f,shieldPower);
            return ShotResult.RESIST;
        }
        else{
            shieldPower = 0.0f;
            return ShotResult.DONOTRESIST;
        }
    }

    public void receiveSupplies(SuppliesPackage s) {
        ammoPower += s.getAmmoPower();
        shieldPower += s.getShieldPower();
        assignFuelValue(fuelUnits + s.getFuelUnits());
    }

    public boolean receiveWeapon(Weapon w) {
        if (hangar == null) {
            return false;
        }

        return hangar.addWeapon(w);
    }

    public Transformation setLoot(Loot loot) {
        CardDealer dealer = CardDealer.getInstance();
        int h = loot.getNHangars();
        if(h > 0){
            Hangar hangarh = dealer.nextHangar();
            this.receiveHangar(hangarh);
        }
        
        int elements = loot.getNSupplies();
        for(int i = 0; i < elements; i++){
            SuppliesPackage sup = dealer.nextSuppliesPackage();
            this.receiveSupplies(sup);
        }
        
        elements = loot.getNWeapons();
        for(int i = 0; i < elements; i++){
            Weapon weap = dealer.nextWeapon();
            this.receiveWeapon(weap);
        }
        
        elements = loot.getNShields();
        for(int i = 0; i < elements; i++){
            ShieldBooster sh = dealer.nextShieldBooster();
            this.receiveShieldBooster(sh);
        }
        int medals = loot.getNMedals();
        nMedals += medals;
        
        if(loot.getEfficient()){
            return Transformation.GETEFFICIENT;
        }
        if(loot.spaceCity()){
            return Transformation.SPACECITY;
        }
        
        return Transformation.NOTRANSFORM;
    }

    public void setPendingDamage(Damage d) {
        pendingDamage = d.adjust(weapons, shieldBoosters);
        cleanPendingDamage();
    }

    public boolean validState() {
        if (pendingDamage == null) {
            return true;
        }
        return pendingDamage.hasNoEffect();
    }

    @Override
    public String toString() {
        String test;

        //CONTENT
        String namE = "+ NAME: " + name + "\n";
        String amP = "+ AMMOPOWER: " + ammoPower + " ";
        String fU = "+ FUELUNITS: " + fuelUnits + " ";
        String sPow = "+ SHIELDPOWER: " + shieldPower + " ";
        String nMed = "+ NMEDALS: " + nMedals + "\n";

        String weap = "\n+ WEAPONS MOUNTED: \n";
        if (weapons == null || weapons.isEmpty()) {
            weap += " Ninguna\n";
        } else {
            for (Weapon w : weapons) {
                weap += w.toString();
            }
        }

        String shB = "\n+ SHIELDBOOSTERS MOUNTED: \n";
        if (shieldBoosters == null || shieldBoosters.isEmpty()) {
            shB += " Ninguno\n";
        } else {
            for (ShieldBooster sB : shieldBoosters) {
                shB += sB.toString();
            }
        }

        String han = "\n* HANGAR: \n";
        if (hangar == null) {
            han += " Ninguno\n";
        } else {
            han += hangar.toString();
        }

        String pDam = "\n* PENDINGDAMAGE: \n";
        if (pendingDamage == null) {
            pDam += " Ninguno\n";
        } else {
            pDam += pendingDamage.toString() + "\n";
        }

        test = namE + amP + fU + sPow + nMed + weap + shB + han + pDam;

        return test;
    }
}